cookbooks/cdo-varnish/libraries/http_cache.rb

```diff
@@ -184,6 +184,7 @@ def self.config(env)
               /v3/animations/*
               /v3/files/*
               /v3/libraries/*
+              /level_starter_assets/*
```

Comment:
Wow. Good find!!! 
I do have one thought: We currently cache many of our Hour of Code tutorials in order to decrease the server load during HOC week. We intentionally strip a lot of information from those in order to prevent a cache bust so the page request never actually hits our server. Will this have any affect on cached levels?

Reply:
(I think James might be the expert in this area if you'd like additional 👀 )

Reply:
From what I understand, this should only affect paths containing `level_starter_assets` so unless HOC assets are using those, I don't believe it should affect them? I'll tag James though to have another look (cc: @jamescodeorg)

Reply:
i built most of the original feature and sanchit and i did some debugging when he started investigating, so i'll try to answer these / give some background, but sanchit can correct me if i'm wrong 😄 

> Are level_starter_assets files user-specific? If not, then I don't think it belongs in this group?

that's a good question/point, james -- level_starter_assets aren't user-specific (just level-specific). authorization comes into play because a levelbuilder can POST and DELETE assets that should be available when a student starts the level. all level_starter_assets for a level are available in the project and show up in the "Manage Assets" modal:
<img width="835" alt="Screen Shot 2021-08-19 at 8 52 33 AM" src="https://user-images.githubusercontent.com/9812299/130101732-b0ee0c38-da65-4a06-b0a2-121e20c4b3d1.png">

> Is the POST coming through with the session cookie? If so, any idea why POST works and DELETE doesn't?

the POST does have the session cookie. the DELETE wasn't working because CanCanCan authorization fails:

https://github.com/code-dot-org/code-dot-org/blob/bc074b9439d1f5cb9217c36f669c495a2eb768dd/dashboard/app/controllers/level_starter_assets_controller.rb#L2

it fails because user/current_user (CanCanCan/Devise terms, respectively) is nil in the DELETE request, which is the bug. we thought it might be because we're skipping the CSRF token check:

https://github.com/code-dot-org/code-dot-org/blob/bc074b9439d1f5cb9217c36f669c495a2eb768dd/dashboard/app/controllers/level_starter_assets_controller.rb#L5

so sanchit tried including the CSRF header with the DELETE request, but i don't think that fixed the issue.

(thought i just had for sanchit -- when you added the CSRF header to the DELETE request, did you also remove the line that skips `verify_authenticity_token`? if not, i think that would be useful to try.)

Reply:
Great find Sanchit! and great question about caching Jessie. This PR would prevent efficient caching of reads to to `/level_starter_assets/*`, which seems risky. If you want to proceed with this, you will have to do some analysis to figure out to convince yourself that this is safe from a scale perspective. 

Alternately, you could fix the editing problems without hurting read performance by using a separate path for editing this data and for serving it at scale. One example of this is how we use /v3/files and /v3/sources for editing (or for reads while a student is editing their own project), but then use /v3/files-public and /v3/sources-public for serving a project in view / share mode.

I hope this helps! please feel free to tag me again if there are other unresolved questions.

Reply:
One note: `/level_starter_assets/*` only gets used on a very small number of levels - all of which (IIRC) are App Lab or Java Lab levels. Many of which are App Lab AI/ML levels.

Reply:
I appreciate all the insight here :) just to add to Maddie's comment (thanks Maddie for the detailed recap!) - 

> it fails because user/current_user (CanCanCan/Devise terms, respectively) is nil in the DELETE request, which is the bug. we thought it might be because we're skipping the CSRF token check:
> 
> https://github.com/code-dot-org/code-dot-org/blob/bc074b9439d1f5cb9217c36f669c495a2eb768dd/dashboard/app/controllers/level_starter_assets_controller.rb#L5

The GET `file` API on level starter assets also has a nil current user for the same reason (no cookies) but since we don't check authenticity for `file` and `show`, we are still able to GET the starter asset file. An alternative fix that doesn't touch caching rules would be disable the authorization check for DELETE in the same way as GET, but that feels dangerous.

https://github.com/code-dot-org/code-dot-org/blob/bc074b9439d1f5cb9217c36f669c495a2eb768dd/dashboard/app/controllers/level_starter_assets_controller.rb#L2

As to why the POST request has the cookie while the DELETE does not - I was suspecting the rules in `http_cache` were treating POST request for static files differently from DELETE requests? Since POST requests don't actually include the filename in the URL vs DELETE and GET do? 

> (thought i just had for sanchit -- when you added the CSRF header to the DELETE request, did you also remove the line that skips verify_authenticity_token? if not, i think that would be useful to try.)

@madelynkasula Hmm, I could try this - I think I did try removing that line but it may have been without the token being sent.

Reply:
> One note: `/level_starter_assets/*` only gets used on a very small number of levels - all of which (IIRC) are App Lab or Java Lab levels. Many of which are App Lab AI/ML levels.

great point Jessie, these are much less common than I had guessed!
![Screen Shot 2021-08-19 at 11 16 27 AM](https://user-images.githubusercontent.com/8001765/130122799-2b6e95d3-a78c-4d97-80fc-e4ae1bbad3af.png)

I've got no objections to breaking efficient caching on reads for this path, then.

In that case, the infra team may want to keep this feature on their radar for future HoC scale analysis in case someone decides to use it in a future HoC script.

Reply:
I think the DELETE request is getting caught in this rule: https://github.com/code-dot-org/code-dot-org/blob/746d5b11a6b879c19dbbc246430e19560aacda1c/cookbooks/cdo-varnish/libraries/http_cache.rb#L232-L237

Reply:
It seems like one part of the problem is that caching behavior can only be defined by path pattern and not by method so we can't define different rules for POST/DELETE than for GET.  Without changing this limitation, it seems like Dave's suggestions (use a different path for POST/DELETE or just don't cache all requests) are the only options.  I'm ok with either.  If we choose to not cache all requests, it would be great if we could split this path out into its own rule, perhaps with a comment explaining why we decided to do this.

Reply:
> I think the DELETE request is getting caught in this rule:
> 
> https://github.com/code-dot-org/code-dot-org/blob/746d5b11a6b879c19dbbc246430e19560aacda1c/cookbooks/cdo-varnish/libraries/http_cache.rb#L232-L237

Yep correct, this is the rule that was stripping cookies for DELETE and GET. For now I'll go ahead and split this into it's own rule to avoid inflating the size of the PR, but will create a follow-up task to explore creating different paths for POST/DELETE and GET.

Reply:
Updated this PR to split this path into its own rule, and creating a follow-up task here: https://codedotorg.atlassian.net/browse/STAR-1689